File: settlement/room/home/chamber/Work.java
package settlement.room.home.chamber;

import static settlement.main.SETT.PATH;
import static settlement.main.SETT.ROOMS;

import game.audio.SoundRace;
import init.race.RACES;
import init.resources.RBIT;
import init.resources.RBIT.RBITImp;
import init.resources.RESOURCE;
import settlement.entity.humanoid.Humanoid;
import settlement.misc.job.SETT_JOB;
import settlement.path.AVAILABILITY;
import settlement.stats.Induvidual;
import settlement.stats.STATS;
import settlement.stats.colls.StatsHome.StatFurniture;
import settlement.stats.equip.WearableResource;
import snake2d.util.datatypes.COORDINATE;
import snake2d.util.datatypes.Coo;
import snake2d.util.misc.CLAMP;

final class Work implements SETT_JOB{
	
	private static final int NEEDS = 0;
	private static final int RESERVED = 1;
	private int data;
	private final Coo coo = new Coo();
	private ChamberInstance ins;
	private final ROOM_CHAMBER blue;
	
	Work(ROOM_CHAMBER blue) {
		this.blue = blue;
	}
	
	Work get(int tx, int ty) {
		if (blue.is(tx, ty)) {
			int data = ROOMS().data.get(tx, ty);
			if (PATH().availability.get(tx, ty)== AVAILABILITY.ROOM) {
				this.data = data;
				this.coo.set(tx, ty);
				this.ins = blue.get(tx, ty);
				
				return this;
			}
		}
		return null;
	}
	
	void clear() {
		ROOMS().data.set(ins, coo, 0);
	}
	
	private void save() {
		
		int old = ROOMS().data.get(coo);
		
		if (old != data) {
			ROOMS().data.set(ins, coo, data);
		}
	}
	
	private int state() {
		return data & 0b01111;
	}
	
	private void stateSet(int state) {
		data &= ~0b01111;
		data |= state;
		save();
	}
	
	@Override
	public boolean jobUseTool() {
		return false;
	}
	
	@Override
	public void jobStartPerforming() {
		
	}
	
	@Override
	public SoundRace jobSound() {
		return ins.blueprintI().employment().sound();
	}
	
	int i = 0;
	
	private final RBITImp bit = new RBITImp();
	
	@Override
	public RBIT jobResourceBitToFetch() {
		i++;
		if ((i & 1) == 0 && !ins.fetching && ins.occupant() != null) {
			bit.clear();
			Induvidual in = ins.occupant().indu();
			for (StatFurniture f : STATS.HOME().getTmp(in)) {
				if (f.needed(ins.occupant().indu()) > 0 && ins.jobs.resourceReachable(f.resource(in)))
					bit.or(f.resource(in));
			}
			
			return bit.isClear() ? null : bit;
		}
		return null;
	}
	
	
	@Override
	public boolean jobReservedIs(RESOURCE r) {
		return state() == RESERVED;
	}
	
	@Override
	public void jobReserveCancel(RESOURCE r) {
		if (jobReservedIs(r)) {
			if (r != null)
				ins.fetching = false;
			stateSet(NEEDS);
		}
	}
	
	@Override
	public boolean jobReserveCanBe() {
		return state() == NEEDS || state() == 2;
	}
	
	@Override
	public void jobReserve(RESOURCE r) {
		if (!jobReserveCanBe())
			throw new RuntimeException();
		stateSet(RESERVED);
		if (r != null)
			ins.fetching = true;
	}
	
	@Override
	public double jobPerformTime(Humanoid skill) {
		return 30;
	}
	
	@Override
	public RESOURCE jobPerform(Humanoid skill, RESOURCE res, int ram) {
		if (!jobReservedIs(res))
			throw new RuntimeException();
		stateSet(NEEDS);
		
		if (res != null) {
			if (ins.occupant() != null) {
				Induvidual in = ins.occupant().indu();
				for (WearableResource rr : RACES.res().get(ins.occupant().indu().popCL(), res)) {
					int nn = rr.needed(in);
					int aa = CLAMP.i(ram, 0, nn);
					rr.inc(in, aa);
					ram -= aa;
					if (ram <= 0)
						break;
				}
				
			}
			ins.fetching = false;
		}
		
		return null;
	}
	
	@Override
	public CharSequence jobName() {
		return ins.blueprintI().employment().verb;
	}
	
	@Override
	public COORDINATE jobCoo() {
		return coo;
	}


	
}

File: LegacySource/settlement/room/home/chamber/Work.cs
using System;
using settlement.main;
using game.audio;
using init.race;
using init.resources;
using settlement.entity.humanoid;
using settlement.misc.job;
using settlement.path;
using settlement.stats;
using settlement.stats.colls;
using settlement.stats.equip;
using snake2d.util.datatypes;
using snake2d.util.misc;

namespace settlement.room.home.chamber
{
    final class Work : SETT_JOB
    {
        private static readonly int NEEDS = 0;
        private static readonly int RESERVED = 1;
        private int data;
        private readonly Coo coo = new Coo();
        private ChamberInstance ins;
        private readonly ROOM_CHAMBER blue;

        public Work(ROOM_CHAMBER blue)
        {
            this.blue = blue;
        }

        public Work Get(int tx, int ty)
        {
            if (blue.Is(tx, ty))
            {
                int data = ROOMS().Data.Get(tx, ty);
                if (PATH().Availability.Get(tx, ty) == AVAILABILITY.ROOM)
                {
                    this.data = data;
                    this.coo.Set(tx, ty);
                    this.ins = blue.Get(tx, ty);

                    return this;
                }
            }
            return null;
        }

        public void Clear()
        {
            ROOMS().Data.Set(ins, coo, 0);
        }

        private void Save()
        {
            int old = ROOMS().Data.Get(coo);

            if (old != data)
            {
                ROOMS().Data.Set(ins, coo, data);
            }
        }

        private int State()
        {
            return data & 0b01111;
        }

        private void StateSet(int state)
        {
            data &= ~0b01111;
            data |= state;
            Save();
        }

        public override bool JobUseTool()
        {
            return false;
        }

        public override void JobStartPerforming()
        {
        }

        public override SoundRace JobSound()
        {
            return ins.BlueprintI().Employment().Sound();
        }

        int i = 0;

        private readonly RBITImp bit = new RBITImp();

        public override RBIT JobResourceBitToFetch()
        {
            i++;
            if ((i & 1) == 0 && !ins.Fetching && ins.Occupant() != null)
            {
                bit.Clear();
                Induvidual inu = ins.Occupant().Indu();
                foreach (StatFurniture f in STATS.HOME().GetTmp(inu))
                {
                    if (f.Needed(ins.Occupant().Indu()) > 0 && ins.Jobs.ResourceReachable(f.Resource(inu)))
                        bit.Or(f.Resource(inu));
                }

                return bit.IsClear() ? null : bit;
            }
            return null;
        }

        public override bool JobReservedIs(RESOURCE r)
        {
            return State() == RESERVED;
        }

        public override void JobReserveCancel(RESOURCE r)
        {
            if (JobReservedIs(r))
            {
                if (r != null)
                    ins.Fetching = false;
                StateSet(NEEDS);
            }
        }

        public override bool JobReserveCanBe()
        {
            return State() == NEEDS || State() == 2;
        }

        public override void JobReserve(RESOURCE r)
        {
            if (!JobReserveCanBe())
                throw new RuntimeException();
            StateSet(RESERVED);
            if (r != null)
                ins.Fetching = true;
        }

        public override double JobPerformTime(Humanoid skill)
        {
            return 30;
        }

        public override RESOURCE JobPerform(Humanoid skill, RESOURCE res, int ram)
        {
            if (!JobReservedIs(res))
                throw new RuntimeException();
            StateSet(NEEDS);

            if (res != null)
            {
                if (ins.Occupant() != null)
                {
                    Induvidual inu = ins.Occupant().Indu();
                    foreach (WearableResource rr in RACES.Res().Get(ins.Occupant().Indu().PopCL(), res))
                    {
                        int nn = rr.Needed(inu);
                        int aa = CLAMP.i(ram, 0, nn);
                        rr.Inc(inu, aa);
                        ram -= aa;
                        if (ram <= 0)
                            break;
                    }

                }
                ins.Fetching = false;
            }

            return null;
        }

        public override CharSequence JobName()
        {
            return ins.BlueprintI().Employment().Verb;
        }

        public override COORDINATE JobCoo()
        {
            return coo;
        }
    }
}